File: BOJ/BOJ/boj 20501 Facebook.py
import sys

input = sys.stdin.readline

N = int(input())
adj = [input() for _ in range(N)]
Q = int(input())

bitset = [[0 for _ in range(N//32 + 1)] for _ in range(N)]
for i in range(N):
    for j in range(N):
        if adj[i][j] == '1':
            bitset[i][j//32] |= (1 << (j % 32))
            
def get_cnt(a,b):
    cnt = 0
    for i in range(32):
        cnt += bin(bitset[a][i] & bitset[b][i]).count('1')
    return cnt

for _ in range(Q):
    a, b = map(int, input().split())
    a -= 1
    b -= 1

    res = 0
    res += get_cnt(a, b)
    print(res)

File: BOJ/BOJ/boj 20501 Facebook.cpp
#include <iostream>
#include <vector>
#include <string>
#include <bitset>

#define BIT_WITDH 32

using namespace std;

typedef long long ll;

bitset<2000> bits[2000];

// int get_cnt(int a, int b){
//     int cnt = 0;
//     for (int i = 0; i < BIT_WITDH; ++i) {
//         cnt += __builtin_popcount(bits[a][i] & bits[b][i]);
//     }
//     return cnt;
// };
int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    cin >> N;

    for (int i = 0; i < N; ++i) {
        string adj;
        cin >> adj;
        bits[i] = bitset<2000>(adj);
    }
    int Q;
    cin >> Q;

    while (Q--) {
        int a, b;
        cin >> a >> b;
        --a, --b;
        int res = (bits[a] & bits[b]).count();
        cout << res << '\n';
    }

    return 0;
}
